tests/unit/test_endpoint_provider.py

```diff
@@ -544,3 +553,141 @@ def test_deeply_nested(self):
             ]
         }
         self.assertDictEqual(expected_rules_file, rules_file_dict_read)
+
+
+def provider_with_rules_factory(rules: Dict[str, str]):
+    """Class factory to create a provider charm class with the given rules.
+
+    Args:
+        rules: a mapping between filename and contents.
+    """
+
+    class ProviderCharm(CharmBase):
+        metadata_yaml = textwrap.dedent(
+            """
+            provides:
+              metrics-endpoint:
+                interface: prometheus_scrape
+            """
+        )
+
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args)
+
+            self.sandbox = TempFolderSandbox()
+            alert_rules_path = os.path.join(self.sandbox.root, "alert_rules")
+            for filepath, contents in rules.items():
+                self.sandbox.put_file(os.path.join(alert_rules_path, filepath), contents)
+
+            self.provider = MetricsEndpointProvider(
+                self, jobs=JOBS, alert_rules_path=alert_rules_path
+            )
+
+    return ProviderCharm
+
+
+class TestAlertRulesDoNotContainUnitNumber(unittest.TestCase):
+    """Feature: Consumer lib must not add a unit number to rule filters or labels.
+
+    Background: In lieu of SDI or pydantic, schema verification is done in unit tests.
+    """
+
+    def setUp(self):
+        pass
+
+    @patch_network_get(private_address="192.0.8.2")
+    def test_alert_rules_with_topology(self):
+        """Scenario: Since only the leader sets rel data, unit number can't be used as a filter."""
+        # GIVEN a provider charm has rule file(s) with a topology stub in the expr
+        provider_charm = provider_with_rules_factory(
+            {
+                "with_topology.rule": yaml.safe_dump(
+                    {"alert": "with_topology", "expr": "up{%%juju_topology%%} == 0"}
+                )
+            }
+        )
+
+        # WHEN relation data is generated
+        self.harness = Harness(provider_charm, meta=provider_charm.metadata_yaml)
```

Comment:
It is unusual to instantiate the Harness in each test of the same `unittest` class. By departing from standard convention we make our tests that bit harder to understand.

Reply:
Doesn't it make sense that unit tests of the Test Harness itself will need instantiate the test harness in the test ? I do not see how this extends to tests that use the Test Harness.

Reply:
Gonna leave this for the OP, @dstathis.